File: linked_list/easy/middle_of_the_linked_list_876.go
package easy

import "github.com/vielendanke/algorithms_in_place/linked_list"

func middleNode(head *linked_list.ListNode) *linked_list.ListNode {
	length := 0

	temp := head

	for temp != nil {
		length++
		temp = temp.Next
	}
	middle := length / 2
	temp = head

	for middle > 0 {
		middle--
		temp = temp.Next
	}
	return temp
}

File: leetcode/linked_list/easy/java_solutions/MiddleOfTheLinkedList_876.java
package leetcode.linked_list.easy.java_solutions;

import static leetcode.linked_list.Helper.ListNode;

public class MiddleOfTheLinkedList_876 {

    static class Solution {
        public ListNode middleNode(ListNode head) {
            ListNode fast = head;
            ListNode slow = head;

            while (fast != null && fast.next != null) {
                fast = fast.next.next;
                slow = slow.next;
            }
            return slow;
        }
    }

    static class SolutionCounter {
        public ListNode middleNode(ListNode head) {
            int n = 0;

            ListNode temp = head;

            while (temp != null) {
                temp = temp.next;
                n++;
            }
            for (int i = 0; i < n / 2; i++) {
                head = head.next;
            }
            return head;
        }
    }
}
